### PR TITLE
Advanced concat for InferenceData

### PR DESCRIPTION
**Pull request by ahartikainen**
If dim=None --> concat unique groups
Valid dim values are {"chain", "draw"}.

~Needs tests.~
~Needs examples.~ (let's update tutorial in another PR)

### CODE REVIEW

**Comment by OriolAbril** [comment]
LGTM.

I also think that in addition to the tutorial, we could add the InferenceData class to the docs. Otherwise, the detailed description of the arguments in `concat` or `sel` cannot be found anywhere. I can create an issue to keep this in mind if you agree.

**Comment by ahartikainen** [comment]
Sounds good.

File: arviz/tests/test_data.py [hunk 1]
```diff
@@ -192,6 +192,58 @@ def test_concat(copy, inplace, sequence):
...
+        new_idata = idata1
+    assert new_idata is not None
+    test_dict = {"posterior": ["A", "B"], "observed_data": ["C", "D"]}
+    fails = check_multiple_attrs(test_dict, new_idata)
+    assert not fails
+    if inplace:
+        assert id(new_idata) == original_idata1_id
+    else:
+        assert id(new_idata) != id(idata1)
+    print(getattr(new_idata.posterior, dim).size)
```
Comment (canyon289) [inline arviz/tests/test_data.py]: Not the biggest of deals. Can this print be changed to a log statement?

File: scripts/create_testenv.sh [hunk 2]
```diff
@@ -56,7 +56,7 @@ if [ "$PYSTAN_VERSION" = "latest" ]; then
 else
   if [ "$PYSTAN_VERSION" = "preview" ]; then
     # try to skip other pre-releases than pystan
-    pip --no-cache-dir install numpy uvloop marshmallow PyYAML
+    pip --no-cache-dir install numpy uvloop marshmallow==3.0.0rc6 PyYAML
```
Comment (canyon289) [inline scripts/create_testenv.sh]: Not asking for a change but curious. What UVloop, and pyYAML used for in pystan? I remember pieces of this conversation in slack but seeing this here am curious.

File: arviz/tests/test_data.py (hunk 1 above)
Comment (ahartikainen) [reply arviz/tests/test_data.py]: There should be no print, it's my debug line

File: scripts/create_testenv.sh (hunk 2 above)
Comment (ahartikainen) [reply scripts/create_testenv.sh]: 5s explanation: They are used by `httpstan` / `pystan3` to do communication and parsing io

https://github.com/stan-dev/httpstan

**Comment by ColCarroll** [comment]
:+1:

**Pull request merged**